sel/src/Sel/SecretKey/AuthenticatedEncryption.hs

```diff
@@ -0,0 +1,388 @@
+  displayBuilder = Builder.fromText . secretKeyToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show SecretKey where
+  show = BS.unpackChars . secretKeyToHexByteString
+
+-- | Generate a new random secret key.
+--
+-- @since 0.0.1.0
+newSecretKey :: IO SecretKey
+newSecretKey = do
+  fPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretboxKeyBytes)
+  Foreign.withForeignPtr fPtr $ \ptr ->
+    cryptoSecretboxKeygen ptr
+  pure $ SecretKey fPtr
+
+-- | Create a 'SecretKey' from a 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- @since 0.0.1.0
+secretKeyFromByteString :: StrictByteString -> SecretKey
+secretKeyFromByteString bytestring = unsafeDupablePerformIO $
+  BS.unsafeUseAsCStringLen bytestring $ \(outsideSecretKeyPtr, _) -> do
+    secretKeyForeignPtr <- BS.mallocByteString @CChar (fromIntegral cryptoSecretboxKeyBytes)
+    Foreign.withForeignPtr secretKeyForeignPtr $ \secretKeyPtr ->
+      Foreign.copyArray outsideSecretKeyPtr secretKeyPtr (fromIntegral cryptoSecretboxKeyBytes)
+    pure $ SecretKey (Foreign.castForeignPtr @CChar @CUChar secretKeyForeignPtr)
+
+-- | Convert a 'SecretKey' to a strict, hexadecimal-encoded 'Text'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToHexText :: SecretKey -> Text
+secretKeyToHexText = Base16.encodeBase16 . secretKeyToBinary
+
+-- | Convert a 'SecretKey' to a strict, hexadecimal-encoded 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToHexByteString :: SecretKey -> StrictByteString
+secretKeyToHexByteString = Base16.encodeBase16' . secretKeyToBinary
+
+-- | Convert a 'SecretKey' to a strict binary 'StrictByteString'.
+--
+-- ⚠️  Be prudent as to where you store it!
+--
+-- @since 0.0.1.0
+secretKeyToBinary :: SecretKey -> StrictByteString
+secretKeyToBinary (SecretKey secretKeyForeignPtr) =
+  BS.fromForeignPtr0
+    (Foreign.castForeignPtr @CUChar @Word8 secretKeyForeignPtr)
+    (fromIntegral @CSize @Int cryptoSecretboxKeyBytes)
+
+-- | A random number that must only be used once per exchanged message.
+-- It does not have to be confidential.
+-- It is of size 'cryptoSecretboxNonceBytes'.
+--
+-- @since 0.0.1.0
+newtype Nonce = Nonce (ForeignPtr CUChar)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq Nonce where
+  (Nonce hk1) == (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretboxKeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord Nonce where
+  compare (Nonce hk1) (Nonce hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretboxKeyBytes
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Display Nonce where
+  displayBuilder = Builder.fromText . nonceToHexText
+
+-- | ⚠️  Be prudent with what you do with it!
+--
+-- @since 0.0.1.0
+instance Show Nonce where
+  show = BS.unpackChars . nonceToHexByteString
```

Comment:
Why would this be a problem if it's not confidential?

Reply:
Because caution is the mother of safety, after all they are essential in making sure replay attacks are impossible

Reply:
Something doesn't add up for me here: on the one hand, our documentation claims that nonces aren't confidential, but on the other, if they can facilitate replays, they sound pretty damn confidential to me.

Reply:
You're right, libsodium does say the nonces don't have to be confidential, but I feel quite paranoid on these matters

Reply:
Paranoia is fine, as long as it's justified. I guess I'm just not clear on what 'not confidential' means. Probably needs an actual expert to weigh in.

Reply:
Nonces need to be unique for each encryption, but can be public, and generated in a predictable way.

Reply:
Thanks! I can now safely withdraw my fears :)